dev: Added recommended settings file

eliotwrobson opened the pull request:
Add recommended vscode config file that automatically runs formatting.

michaelfromyeg commented:
There's a couple of other Python extensions we could recommend; this is what I have from another project.

```
    "ms-python.python",
    "ms-python.vscode-pylance",
    "ms-python.pylint", // should be skipped; same as flake8
    "ms-python.mypy-type-checker",
    "ms-python.black-formatter",
    "ms-python.isort", // definitely worth adding here!
```

Plus maybe `"tamasfe.even-better-toml"` for a `pyproject.toml`.

Or as a bonus, `"EditorConfig.EditorConfig"` for EditorConfig support.

michaelfromyeg commented:
In terms of the `settings.json`, I'm pro bringing this into source control vs. tucked away `README.md`; it's just so nice have this stuff behave nicely out of the box (since, iirc, it's enforced in CI anyway) and we don't have `pre-commit` or similar, as discussed on Slack.

That being said, I see @jonatanschroeder's point about wiping out folks' existing settings. Do we have a nice way to address this? (Is a #pl-dev @channel sufficient?)

nwalters512 commented:
https://github.com/microsoft/vscode/issues/40233 and https://github.com/microsoft/vscode/issues/15909 talk about a lot of the pitfalls of this approach, with [this comment](https://github.com/microsoft/vscode/issues/40233#issuecomment-1259452643) summing things up best IMO:

> what makes this a pain point is that the moment it becomes necessary to track any project configuration in source control, developers must give up local control over all settings defined in settings.json.

At this moment, I'm opposed to this change for exactly this reason.

eliotwrobson commented:
Closing for reasons outlined in Nate's last comment. We can always point people to this PR to grab the settings file (and mention something in the docs).

nwalters512 commented:
As a stopgap, I'd be happy to literally write a page in the docs with our recommended editor setup and/or commit a `.vscode/settings.example.json` file.

nwalters512 commented:
I'd also be more than happy to approve just the `.vscode/extensions.json` file from this diff.

nwalters512 commented:
A final interesting bit of trivia: the VSCode team themselves commit `.vscode/settings.json` 🙃

eliotwrobson commented:
> As a stopgap, I'd be happy to literally write a page in the docs with our recommended editor setup and/or commit a `.vscode/settings.example.json` file.

I think this is the best solution. I think adding the recommended extensions helps, but the real issue is that to get formatting-on-save, you have to set that up in `settings.json` using the weird custom stuff we have in that file. So putting this in the docs, any new contributor is going to have to set that up for themselves.

> A final interesting bit of trivia: the VSCode team themselves commit `.vscode/settings.json` 🙃

At some level, it makes sense. Unless developers are doing weird custom things with their editors, most people are going to be roughly as productive with the same set of editor behavior.

nwalters512 commented:
The recommended extensions would at least poke people to set them up when they first open, and there's a nice UI for viewing/installing the recommended ones, so I'd still like to make that change.

We can also document the need to install Python deps natively, in case one is using Docker for dev.

eliotwrobson commented:
Removed settings and updated extensions, issue for adding recommended settings to docs is here: https://github.com/PrairieLearn/PrairieLearn/issues/8279

nwalters512 reviewed (approved):
I would've opened a new PR and left this one closed as a record of what we didn't do.
